# Project 1 : Titanic Survive Analysis

In [3]:
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn.preprocessing import LabelEncoder

In [48]:
titanic_train = pd.read_csv("train.csv")

In [49]:
titanic_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [50]:
titanic_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Name         889 non-null    object 
 4   Sex          889 non-null    object 
 5   Age          889 non-null    float64
 6   SibSp        889 non-null    int64  
 7   Parch        889 non-null    int64  
 8   Ticket       889 non-null    object 
 9   Fare         889 non-null    float64
 10  Cabin        202 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.5+ KB


In [51]:
titanic_train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

### Training the Model {Classification}

In [52]:
label_encoder = LabelEncoder()
encoded_sex = label_encoder.fit_transform(titanic_train["Sex"])

In [53]:
predictors = pd.DataFrame([encoded_sex,titanic_train["Age"],titanic_train["Fare"]]).T

In [54]:
tree_model = tree.DecisionTreeClassifier()

In [55]:
tree_model.fit(X = predictors,y = titanic_train["Survived"])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [56]:
with open ("Tree_Model.dot","w") as g:
    g = tree.export_graphviz(tree_model,feature_names =["Sex","Age","Fare"],out_file = g);

In [57]:
tree_model.score(X = predictors,y=titanic_train["Survived"])

0.9763779527559056

### Testing the model {Prediction}

In [58]:
titanic_test = pd.read_csv("test.csv")

In [59]:
titanic_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Embarked  
0  34.5      0      0   330911   7.8292        Q  
1  47.0      1      0   363272   7.0000        S  
2  62.0      0      0   240276   9.6875        Q  
3  27.0      0      0   315154   8.6625        S  
4  22.0      1      1  3101298  12.2875        S

In [60]:
titanic_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 417 entries, 0 to 416
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  417 non-null    int64  
 1   Pclass       417 non-null    int64  
 2   Name         417 non-null    object 
 3   Sex          417 non-null    object 
 4   Age          417 non-null    float64
 5   SibSp        417 non-null    int64  
 6   Parch        417 non-null    int64  
 7   Ticket       417 non-null    object 
 8   Fare         417 non-null    float64
 9   Embarked     417 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 32.7+ KB


In [61]:
encoded_sex1 = label_encoder.fit_transform(titanic_test["Sex"])

In [62]:
predictors1 = pd.DataFrame([encoded_sex1,titanic_test["Age"],titanic_test["Fare"]]).T

In [63]:
model_predict = tree_model.predict(X = predictors1)

In [64]:
result = pd.DataFrame({"PassengerId":titanic_test["PassengerId"],"Survived":model_predict})

In [65]:
result.to_csv("Titanic_Prediction.csv",index = False)

In [66]:
check = pd.read_csv("Titanic_Prediction.csv")

In [67]:
check.head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         1
4          896         0

# Project 2 : Attrition Analysis 

In [68]:
attrition = pd.read_csv("general_data.csv")

In [71]:
attrition.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [72]:
attrition.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [73]:
#Function for Lable Encoding

def encode(cols):
    enc = LabelEncoder()
    for col in cols:
        attrition[col] = enc.fit_transform(attrition[col])

In [74]:
encode_cols = ['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus']
encode(encode_cols)

In [75]:
attrition.head()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0   51          0               2           2                 6          2   
1   31          1               1           1                10          1   
2   32          0               1           1                17          4   
3   38          0               0           1                 2          5   
4   32          0               2           1                10          1   

   EducationField  EmployeeCount  EmployeeID  Gender  ...  NumCompaniesWorked  \
0               1              1           1       0  ...                 1.0   
1               1              1           2       0  ...                 0.0   
2               4              1           3       1  ...                 1.0   
3               1              1           4       1  ...                 3.0   
4               3              1           5       1  ...                 4.0   

   Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
0       Y                 11              8                 0   
1       Y                 23              8                 1   
2       Y                 15              8                 3   
3       Y                 11              8                 3   
4       Y                 12              8                 2   

  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0               1.0                      6               1   
1               6.0                      3               5   
2               5.0                      2               5   
3              13.0                      5               8   
4               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

### Using Random Forest Algorithm to find the Important IDV

In [76]:
from sklearn.ensemble import RandomForestClassifier

In [78]:
random_forest = RandomForestClassifier(n_estimators=1000, max_features=2,oob_score=True)

In [121]:
features = ["Age","BusinessTravel","Department","Education","EducationField","Gender","JobLevel","JobRole",\
           "MaritalStatus","StockOptionLevel"]

In [122]:
random_forest.fit(X = attrition[features],y=attrition["Attrition"])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [123]:
print(random_forest.oob_score_)

0.998639455782313


In [124]:
for features,imp in zip (features,random_forest.feature_importances_):
    print (features,imp)

Age 0.27485681877187557
BusinessTravel 0.049913708962248345
Department 0.050623438546253285
Education 0.10186680400708875
EducationField 0.09250805269086833
Gender 0.03937705172399124
JobLevel 0.09671628857545965
JobRole 0.14336101442150243
MaritalStatus 0.066700906756678
StockOptionLevel 0.08407591554403444


___As per the above list the important features are___

1. *Age*
2. *Education*
3. *Job Role*

### Using Decision Tree to build the model based on the important IDV

In [97]:
predictors3 = pd.DataFrame([attrition["Age"],attrition["Education"],attrition["JobRole"]]).T

In [134]:
attrition_model = tree.DecisionTreeClassifier(max_depth = 6)

In [135]:
attrition_model.fit(X = predictors3,y = attrition["Attrition"])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=6, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [138]:
with open("Attrition_Tree.dot","w") as g:
    g = tree.export_graphviz(attrition_model,feature_names =["Age","Education","JobRole"],out_file = g);

In [139]:
attrition_model.score(X = predictors3, y = attrition["Attrition"])

0.8510204081632653

#### Rules corresponding to the generated Tree leaf nodes

1. Age is less than 21.5 and Education is less than 1.5, Attrition is NO {3 records}
2. Age is more than 58.5, Attrition is NO {45 records}
3. Age is more than 53.5 Attrition is NO {54 records}
4. Age is less than 57.5 and Education less than 1.5, Attrition is NO {18 records}
5. Age is more than 57.5, education is more than 2.5 and Jobrole is less than 5, Attrition is NO {12 records}
6. Age is less than 30.5 and Jobrole is less than 5.5, Attrition is NO {63 records}
7. Age is less than 30.5 and Education is less than 3.5, Attrition is NO {6 records}
8. Age is more than 25, Attrition is NO {93 records}
9. Age is less than 25 and education is more than 2.5, Attrition is NO {15 records}
10. Jobrole is more than 0.5 and Age is less than 20.5, Attrition is NO {15 records}
11. Education is more than 1.5, Age is less than 18.5 and Jobrole is less than 6.5, Attrition is NO {3 records}
-----------------------------------------------------------------------------------------------------------------------------
1. Age is more than 57.5 and education is less than 2.5 Attrition is YES {9 records}
2. . Age is less than 25 and education is less than 2.5, Attrition is YES {3 records}
3. Age is more than 18.5 and Jobrole is less than 0.5, Attrition is YES {3 records}
4. Age is less than 18.5 and Education is more than 4.5, Attrition is YES {3 records}
5. Age is less than 18.5, Education is less than 4.5 and Jobrole is less than 2.5, Attrition is YES {3 records}
6. Education is more than 1.5, Age is more than 18.5 and Jobrole is less than 1, Attrition is YES {9 records}


# Project 3 : Bank Loan Model

In [142]:
bank_loan = pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name = 1)

In [143]:
bank_loan.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [144]:
bank_loan.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [146]:
bank_loan = bank_loan.drop("ZIP Code", axis = 1)

### Using Random Forest Algorithm to find the Important IDV

In [148]:
rf_model = RandomForestClassifier(n_estimators=1000, max_features=2, oob_score=True)

In [149]:
features1 = ["Age","Experience","Income","Family","CCAvg","Education","Mortgage","Securities Account",\
           "CD Account","Online","CreditCard"]

In [151]:
rf_model.fit(X = bank_loan[features1],y=bank_loan["Personal Loan"])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [152]:
print(rf_model.oob_score_)

0.9876


In [153]:
for features1,imp in zip (features1,rf_model.feature_importances_):
    print (features1,imp)

Age 0.045003296580786896
Experience 0.04462867729323726
Income 0.3468683708460796
Family 0.0952900343127949
CCAvg 0.18128916354894953
Education 0.16309822261415247
Mortgage 0.04400181723753674
Securities Account 0.005538399114860752
CD Account 0.05615663536927635
Online 0.008336945718877567
CreditCard 0.009788437363447934


___As per the above results the important IDV are___

1. *Income*
2. *CCAvg*
3. *Education*

### Using Decision Tree to build the model based on the important IDV

In [154]:
predictors4 = pd.DataFrame([bank_loan["Income"],bank_loan["CCAvg"],bank_loan["Education"]]).T

In [160]:
bank_model = tree.DecisionTreeClassifier(max_depth = 6)

In [161]:
bank_model.fit(X = predictors4,y = bank_loan["Personal Loan"])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=6, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [162]:
with open("BankLoan_Tree.dot","w") as g:
    g = tree.export_graphviz(bank_model,feature_names =["Income","CCAvg","Education"],out_file = g);

In [164]:
bank_model.score(X = predictors4, y = bank_loan["Personal Loan"])

0.9738

#### Rules corresponding to the generated Tree leaf nodes

1. Education is more than 1.5 and Income is more than 116.5, Bank loan Sanctioned is YES {314 Records}
2. CCAvg is more than 4.0, Bank loan Sanctioned is YES {4 Records}
3. CCAvg is less than 0.55, Bank loan Sanctioned is YES {1 record}
4. CCAvg is more than 0.55 and Income is less than 114.5, Bank loan Sanctioned is NO {7 Records}
5. CCAvg is more than 6.55, Bank loan Sanctioned is YES {2 Records}
6. Income is less than 176.5 and CCAvg is more than 6.95, Bank loan Sanctioned is NO {74 Records}
7. Income is less than 178.5 and CCAvg is more than 7.6, Bank loan Sanctioned is YES {1 Record}
8. Income is less than 178.5 and CCAvg is less than 7.6, Bank loan Sanctioned is NO {1 Record}
9. Income is more than 178.5 and CCAvg is more than 8.05, Bank loan Sanctioned is NO {19 Records}
10. CCAvg less than 6.85, Bank loan Sanctioned is NO {15 Records}
11. CCAvg is more than 4.05 and Income is more than 179.5, Bank loan Sanctioned is NO {25 Records}
12. Education is more than 1.5 and income is more than 109, Bank loan Sanctioned is YES {8 Records}
13. Education is less than 1.5 and CCAvg more than 4.65, Bank loan Sanctioned is NO {28 Records}
14. CCAvg is less than 2.95 and income is less than 106.5, Bank loan Sanctioned is NO {3629 Records}